delta-app/src/main/java/io/cdap/delta/app/metrics/MetricsHandler.java

```diff
@@ -0,0 +1,231 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.delta.app.metrics;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import io.cdap.cdap.api.metrics.Metrics;
+import io.cdap.delta.api.DDLOperation;
+import io.cdap.delta.api.DMLOperation;
+import io.cdap.delta.api.SourceTable;
+import io.cdap.delta.app.DeltaWorkerId;
+import io.cdap.delta.app.EventMetrics;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Aggregate event count by table and emit/log metrics
+ */
+public class MetricsHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsHandler.class);
+  private static final String PROGRAM_METRIC_ENTITY = "ent";
+  private static final String DOT_SEPARATOR = ".";
+  private static final String DEFAULT_AGGR_STATS_INTERVAL_SEC = "600"; // 10 min
+  static final String AGGREGATE_STATS_FREQUENCY_ARG = "aggregate.stats.frequency.sec";
+
+  private final Metrics metrics;
+  private final Map<String, EventMetrics> consumerTableEventMetrics;
+  private final Map<String, AtomicReference<EventCounts>> tablePublishedEventCounts;
+  private final Map<String, AtomicReference<EventCounts>> tableConsumedEventCounts;
+  private final int logAggregateStatsIntervalSec;
+  private final ScheduledExecutorService aggregateStatsExecutor;
+
+  public MetricsHandler(DeltaWorkerId id, Metrics metrics, List<SourceTable> tables,
+                        Map<String, String> runtimeArguments) {
+    this.metrics = metrics;
+    this.consumerTableEventMetrics = new HashMap<>();
+    //HashMap is fine as we are not structurally modifying the map after initialization
+    this.tablePublishedEventCounts = new HashMap<>();
+    this.tableConsumedEventCounts = new HashMap<>();
+    tables.forEach(table -> {
+      String fullyQualifiedTableName =
+        getFullyQualifiedTableName(table.getDatabase(), table.getSchema(), table.getTable());
+      tablePublishedEventCounts.put(fullyQualifiedTableName, new AtomicReference<>(new EventCounts()));
+      tableConsumedEventCounts.put(fullyQualifiedTableName, new AtomicReference<>(new EventCounts()));
+    });
+
+    String prefix = String.format("metrics-deltaworker-%d", id.getInstanceId());
+    this.aggregateStatsExecutor = Executors.newScheduledThreadPool(1,
```

Comment:
Why not use newSingleThreadExecutor instead ?

Reply:
Because we need scheduled execution of task at certain interval